mindsdb_native/libs/backends/lightwood.py

```diff
@@ -278,38 +278,95 @@ def train(self):
 
+        lightwood_train_ds = lightwood.api.data_source.DataSource(
+            train_df,
+            config=lightwood_config
+        )
+        lightwood_test_ds = lightwood_train_ds.make_child(test_df)
 
-        self.predictor = lightwood.Predictor(lightwood_config)
+        self.transaction.lmd['lightwood_data']['save_path'] = os.path.join(CONFIG.MINDSDB_STORAGE_PATH, self.transaction.lmd['name'], 'lightwood_data')
+        Path(CONFIG.MINDSDB_STORAGE_PATH).joinpath(self.transaction.lmd['name']).mkdir(mode=0o777, exist_ok=True, parents=True)
 
-        # Evaluate less often for larger datasets and vice-versa
-        eval_every_x_epochs = int(round(1 * pow(10, 6) * (1 / len(train_df))))
+        logging.getLogger().setLevel(logging.DEBUG)
 
-        # Within some limits
-        if eval_every_x_epochs > 200:
-            eval_every_x_epochs = 200
-        if eval_every_x_epochs < 3:
-            eval_every_x_epochs = 3
+        predictors_and_accuracies = []
 
-        if lightwood_config['mixer']['class'] == NnMixer:
-            lightwood_config['mixer']['kwargs']['callback_on_iter'] = self.callback_on_iter
-            lightwood_config['mixer']['kwargs']['eval_every_x_epochs'] = eval_every_x_epochs
+        if self.transaction.lmd['mixer_class'] is not None:
+            mixer_classes = [self.transaction.lmd['mixer_class']]
+        else:
+            mixer_classes = lightwood.mixers.BaseMixer.__subclasses__()
 
-            if self.transaction.lmd['stop_training_in_x_seconds'] is not None:
-                lightwood_config['mixer']['kwargs']['stop_training_after_seconds'] = self.transaction.lmd['stop_training_in_x_seconds']
+        for mixer_class in mixer_classes:
+            lightwood_config['mixer']['kwargs'] = {}
+            lightwood_config['mixer']['class'] = mixer_class
 
-        logging.getLogger().setLevel(logging.DEBUG)
+            if lightwood_config['mixer']['class'] == lightwood.mixers.NnMixer:
+                # Evaluate less often for larger datasets and vice-versa
+                eval_every_x_epochs = int(round(1 * pow(10, 6) * (1 / len(train_df))))
 
-        self.predictor.learn(from_data=train_df, test_data=test_df)
+                # Within some limits
+                if eval_every_x_epochs > 200:
+                    eval_every_x_epochs = 200
+                if eval_every_x_epochs < 3:
+                    eval_every_x_epochs = 3
 
-        self.transaction.log.info('Training accuracy of: {}'.format(self.predictor.train_accuracy))
+                kwargs = lightwood_config['mixer']['kwargs']
+
+                kwargs['callback_on_iter'] = self.callback_on_iter
+                kwargs['eval_every_x_epochs'] = eval_every_x_epochs
+                kwargs['stop_training_after_seconds'] = self.transaction.lmd['stop_training_in_x_seconds']
+
+            self.predictor = lightwood.Predictor(lightwood_config.copy())
+
+            self.predictor.learn(
```

Comment:
I think this will suffer from the issue of encoders being re-trained, we should look at encoder transfer before deploying this (see issues lightwood side), unless that's already implemented and I'm wrong here ?

Alternative here might be to just:

1. Deepcopy predictor
2. Change mixer
3. Make sure lightwood keeps the old encoders which are all marked as "prepared"

Reply:
data source is only created once: https://github.com/mindsdb/mindsdb_native/pull/191/files/d102fdec763ac90b647fbfb7b8ce07adc9a81d4b#diff-ab4b77a7fb52eb86c1fb66a4cbe8b6abR281

so encoders are trained once in `DataSource.__init__`